File: server/app/simulation/race_engine/race_field.py
"""Multi-horse field coordination -- NOT part of the original TS engine
(which runs every RaceSolver in isolation, oblivious to other horses). This
is the foundation for Global-accuracy mechanics that genuinely depend on
where other horses are: live order/rank, gap-to-ahead/behind (bashin_diff_*,
distance_diff_top, order/order_rate), position-keep relative to a real
pacemaker, blocking, and lane selection.

`RaceSolver.field` is set to a `RaceField` once the whole roster is built
(race_runner.py does this after constructing every solver, before the tick
loop starts). It is `None` during the very first "gate skill" activation
round inside `RaceSolver.__init__` -- at that point no other horse exists
yet -- so every dynamic condition that reads `s.field` must handle `None`
by treating the condition as not-yet-determinable (return False).
"""

from __future__ import annotations

from typing import Optional

from .horse_types import Strategy, strategy_matches


class RaceField:
    def __init__(self, solvers: list):
        self.solvers = solvers
        self.n = len(solvers)
        # index -> 1-based rank (1 = leader), updated after every full tick
        self.rank_of: list[int] = [0] * self.n
        # ranks sorted ascending -> solver index, i.e. order[0] is the leader
        self.order: list[int] = list(range(self.n))
        # previous tick's rank, for overtake/order-improvement detection
        self._prev_rank_of: list[int] = list(self.rank_of)
        for s in solvers:
            s.field = self
        self.update()

    def update(self) -> None:
        """Recompute live order/rank from each solver's current `pos`. Call
        once after every horse has been stepped for this tick."""
        self._prev_rank_of = self.rank_of
        idx = sorted(range(self.n), key=lambda i: -self.solvers[i].pos)
        self.order = idx
        rank_of = [0] * self.n
        for rank, i in enumerate(idx):
            rank_of[i] = rank + 1
        self.rank_of = rank_of

    def _index_of(self, solver) -> int:
        # n is at most 18 (a full gate); linear scan is cheap and avoids
        # solvers needing to know their own field-assigned index.
        for i, s in enumerate(self.solvers):
            if s is solver:
                return i
        raise ValueError("solver not in this field")

    def order_of(self, solver) -> int:
        """1-based live finishing-position rank (1 = leader)."""
        return self.rank_of[self._index_of(solver)]

    def previous_order_of(self, solver) -> int:
        return self._prev_rank_of[self._index_of(solver)]

    def leader_pos(self) -> float:
        return self.solvers[self.order[0]].pos

    def distance_to_leader(self, solver) -> float:
        """Meters behind the current leader (0 if this solver IS the leader)."""
        return max(0.0, self.leader_pos() - solver.pos)

    def distance_to_ahead(self, solver) -> Optional[float]:
        """Meters behind the horse immediately ahead, or None if leading."""
        i = self._index_of(solver)
        r = self.rank_of[i]
        if r <= 1:
            return None
        ahead = self.order[r - 2]
        return max(0.0, self.solvers[ahead].pos - solver.pos)

    def distance_to_behind(self, solver) -> Optional[float]:
        """Meters ahead of the horse immediately behind, or None if last."""
        i = self._index_of(solver)
        r = self.rank_of[i]
        if r >= self.n:
            return None
        behind = self.order[r]
        return max(0.0, solver.pos - self.solvers[behind].pos)

    def lane_gap_to_ahead(self, solver) -> Optional[float]:
        i = self._index_of(solver)
        r = self.rank_of[i]
        if r <= 1:
            return None
        ahead = self.order[r - 2]
        return self.solvers[ahead].lane - solver.lane

    def lane_gap_to_behind(self, solver) -> Optional[float]:
        i = self._index_of(solver)
        r = self.rank_of[i]
        if r >= self.n:
            return None
        behind = self.order[r]
        return solver.lane - self.solvers[behind].lane

    def did_overtake(self, solver) -> bool:
        """True on the tick this solver's rank improved (took over at least
        one other horse) versus the previous tick."""
        i = self._index_of(solver)
        return self.rank_of[i] < self._prev_rank_of[i]

    # -- lane / blocking / vision (doc's Lane System + Blocking sections) ---
    # 1 horse-lane = 1/18 course-width (doc). `solver.lane` is stored in
    # course-width units throughout, matching the doc's primary unit; this
    # constant converts a horse-lane count into that unit where the doc
    # states a threshold in horse-lanes.
    HORSE_LANE = 1.0 / 18.0

    def _others(self, solver):
        for s in self.solvers:
            if s is not solver:
                yield s

    def front_blocker(self, solver):
        """Doc: front block = 0 < distGap < 2m AND |laneGap| <= a tolerance
        that shrinks from 0.75 horse-lanes at 0m to 0.3 at 2m. Lowest
        distance-gap qualifying horse wins. Returns (blocker, dist_gap) or
        None."""
        best = None
        for other in self._others(solver):
            dist_gap = other.pos - solver.pos
            if not (0 < dist_gap < 2.0):
                continue
            lane_gap = abs(other.lane - solver.lane)
            tolerance = (1.0 - 0.6 * dist_gap / 2.0) * 0.75 * self.HORSE_LANE
            if lane_gap <= tolerance and (best is None or dist_gap < best[1]):
                best = (other, dist_gap)
        return best

    def side_room(self, solver):
        """Doc: side block = |distGap| < 1.05m AND |laneGap| < 2 horse-lanes.
        Returns (inner_blocked, outer_blocked) -- inner = lower lane (toward
        the rail), outer = higher lane."""
        inner_blocked = outer_blocked = False
        for other in self._others(solver):
            if abs(other.pos - solver.pos) >= 1.05:
                continue
            lane_gap = other.lane - solver.lane
            if abs(lane_gap) >= 2.0 * self.HORSE_LANE:
                continue
            if lane_gap < 0:
                inner_blocked = True
            elif lane_gap > 0:
                outer_blocked = True
        return inner_blocked, outer_blocked

    def is_surrounded(self, solver) -> bool:
        """Doc: true if ALL THREE of Out/Front/Behind each find another
        horse (the same horse may satisfy more than one direction)."""
        out = front = behind = False
        for other in self._others(solver):
            dist_gap = other.pos - solver.pos
            lane_gap = other.lane - solver.lane
            if abs(dist_gap) < 1.5 and 0 < lane_gap < 3.0 * self.HORSE_LANE:
                out = True
            if 0 < dist_gap < 3.0 and abs(lane_gap) < 1.5 * self.HORSE_LANE:
                front = True
            if -3.0 < dist_gap < 0 and abs(lane_gap) < 1.5 * self.HORSE_LANE:
                behind = True
        return out and front and behind

    def is_solo_front_runner(self, solver) -> bool:
        return not any(strategy_matches(other.horse.strategy, Strategy.NIGE) for other in self._others(solver))

    def pacemaker(self):
        """Doc's pre-1.5-anniversary pacemaker rule (the current, post-1.5,
        mechanism is explicitly flagged as unclear even in the real game --
        using the last confirmed one): the 1st-place horse among the
        field's most-forward strategy group that's actually present."""
        groups = (
            lambda strat: strat in (Strategy.NIGE, Strategy.OONIGE),
            lambda strat: strat == Strategy.SENKOU,
            lambda strat: strat == Strategy.SASI,
            lambda strat: strat == Strategy.OIKOMI,
        )
        for match in groups:
            for i in self.order:  # self.order is sorted leader-first
                if match(self.solvers[i].horse.strategy):
                    return self.solvers[i]
        return None

    def gap_to_pacemaker(self, solver) -> float:
        pm = self.pacemaker()
        if pm is None or pm is solver:
            return 0.0
        return pm.pos - solver.pos

    def near_count(self, solver) -> int:
        """Doc: near = |distGap| < 3m AND |laneGap| < 3 horse-lanes
        (post-1st-anniversary thresholds; the doc notes these were tighter,
        1.5m/1.5 horse-lanes, before that)."""
        count = 0
        for other in self._others(solver):
            if abs(other.pos - solver.pos) < 3.0 and abs(other.lane - solver.lane) < 3.0 * self.HORSE_LANE:
                count += 1
        return count

    def near_infront_count(self, solver) -> int:
        """Doc: near_infront_count -- same "near" definition as near_count
        (|distGap|<3m, |laneGap|<3 horse-lanes) but restricted to horses
        AHEAD (distGap>0)."""
        count = 0
        for other in self._others(solver):
            dist_gap = other.pos - solver.pos
            if 0 < dist_gap < 3.0 and abs(other.lane - solver.lane) < 3.0 * self.HORSE_LANE:
                count += 1
        return count

    def temptation_count_behind(self, solver) -> int:
        """Doc: temptation_count_behind -- the number of horses ranked
        behind `solver` (by finishing-position rank, not physical distance)
        that are currently rushing (kakari)."""
        r = self.rank_of[self._index_of(solver)]
        return sum(1 for i in self.order[r:] if self.solvers[i].is_kakari)

    def temptation_count_infront(self, solver) -> int:
        """Doc: temptation_count_infront -- the number of horses ranked
        ahead of `solver` that are currently rushing (kakari)."""
        r = self.rank_of[self._index_of(solver)]
        return sum(1 for i in self.order[:r - 1] if self.solvers[i].is_kakari)

    def running_style_temptation_count(self, strategy) -> int:
        """Doc: running_style_temptation_count_{nige,senko,sashi,oikomi} --
        the number of horses in the field (the querying horse counts too)
        with the given running strategy that are currently rushing."""
        return sum(1 for s in self.solvers if strategy_matches(s.horse.strategy, strategy) and s.is_kakari)

    def step_all(self, dt: float, active_indices: Optional[list[int]] = None) -> None:
        """Step every (or only the still-racing `active_indices`) solver by
        `dt`, then refresh the shared order/rank snapshot. Solvers must all
        be stepped before `update()` runs so the snapshot reflects a single
        consistent instant, matching how the original per-horse lockstep
        loop in race_runner.py already worked for frame recording."""
        indices = active_indices if active_indices is not None else range(self.n)
        for i in indices:
            self.solvers[i].step(dt)
        self.update()

File: server/race_engine_rs/src/race_field.rs
//! Port of race_field.py -- multi-horse field coordination. Not part of the
//! original TS engine (which runs every RaceSolver in isolation, oblivious
//! to other horses).
//!
//! `RaceField` owns every solver in `RefCell`s so one can be borrowed
//! mutably (to step it) while its own methods borrow the others
//! immutably to answer blocking/lane/order queries -- the same access
//! pattern Python gets for free via shared references. Solvers do not hold
//! a reference back to the field: instead `field: Option<&RaceField>` is
//! threaded as an explicit parameter through `RaceSolver::step` and every
//! condition-checking closure (see `activation_conditions::DynCond`), so
//! nothing here needs a self-referential pointer or unsafe code.
//!
//! Ported already carrying the perf fixes validated on the Python engine:
//! each solver's own roster index is stored on it directly (`field_index`)
//! rather than found by a linear scan, `side_room`/`is_surrounded` early-out
//! once both/all flags are set, and `pacemaker()` is computed once per tick
//! and cached rather than recomputed per asking horse.

use std::cell::{Cell, RefCell};

use crate::horse_types::{strategy_matches, Strategy};
use crate::race_solver::RaceSolver;

/// 1 horse-lane = 1/18 course-width (doc). `lane` fields elsewhere are
/// stored in course-width units; this converts a horse-lane count into
/// that unit wherever the doc states a threshold in horse-lanes.
pub const HORSE_LANE: f64 = 1.0 / 18.0;

/// Doc's pacemaker strategy groups, most-forward first.
const PACEMAKER_GROUPS: &[&[Strategy]] = &[&[Strategy::Nige, Strategy::Oonige], &[Strategy::Senkou], &[Strategy::Sasi], &[Strategy::Oikomi]];

pub struct RaceField {
    pub solvers: Vec<RefCell<RaceSolver>>,
    pub n: usize,
    /// Each solver's fixed strategy, mirrored outside its `RefCell` so
    /// `pacemaker()` can scan every horse's strategy -- including the one
    /// currently mid-`step()` and thus already mutably borrowed -- without
    /// re-entering that `RefCell`.
    strategies: Vec<Strategy>,
    rank_of: RefCell<Vec<usize>>,
    order: RefCell<Vec<usize>>,
    prev_rank_of: RefCell<Vec<usize>>,
    pacemaker_idx: Cell<Option<usize>>,
    pacemaker_valid: Cell<bool>,
}

impl RaceField {
    pub fn new(mut solvers: Vec<RaceSolver>) -> RaceField {
        let n = solvers.len();
        for (i, s) in solvers.iter_mut().enumerate() {
            s.field_index = Some(i);
        }
        let strategies = solvers.iter().map(|s| s.horse.strategy).collect();
        let field = RaceField {
            solvers: solvers.into_iter().map(RefCell::new).collect(),
            n,
            strategies,
            rank_of: RefCell::new(vec![0; n]),
            order: RefCell::new((0..n).collect()),
            prev_rank_of: RefCell::new(vec![0; n]),
            pacemaker_idx: Cell::new(None),
            pacemaker_valid: Cell::new(false),
        };
        field.update();
        field
    }

    /// Recompute live order/rank from each solver's current `pos`. Call
    /// once after every horse has been stepped for this tick.
    pub fn update(&self) {
        let mut prev = self.prev_rank_of.borrow_mut();
        let rank = self.rank_of.borrow();
        prev.copy_from_slice(&rank);
        drop(rank);
        self.pacemaker_valid.set(false);

        let positions: Vec<f64> = self.solvers.iter().map(|s| s.borrow().pos).collect();
        let mut idx: Vec<usize> = (0..self.n).collect();
        idx.sort_by(|&a, &b| positions[b].partial_cmp(&positions[a]).unwrap());
        let mut rank_of = vec![0usize; self.n];
        for (rank, &i) in idx.iter().enumerate() {
            rank_of[i] = rank + 1;
        }
        *self.rank_of.borrow_mut() = rank_of;
        *self.order.borrow_mut() = idx;
    }

    /// Step every (or only the still-racing `active_indices`) solver by
    /// `dt`, then refresh the shared order/rank snapshot. Solvers must all
    /// be stepped before `update()` runs so the snapshot reflects a single
    /// consistent instant.
    pub fn step_all(&self, dt: f64, active_indices: &[usize]) {
        for &i in active_indices {
            self.solvers[i].borrow_mut().step(dt, Some(self));
        }
        self.update();
    }

    pub fn order_of(&self, i: usize) -> usize {
        self.rank_of.borrow()[i]
    }

    pub fn previous_order_of(&self, i: usize) -> usize {
        self.prev_rank_of.borrow()[i]
    }

    /// `my_pos`/`my_lane` below are the querying horse `i`'s own current
    /// values, passed in by the caller instead of re-borrowed here -- `i`
    /// may be the solver currently mid-`step()` and thus already mutably
    /// borrowed via its `RefCell`, so this must never index `self.solvers[i]`.
    pub fn distance_to_leader(&self, i: usize, my_pos: f64) -> f64 {
        let leader = self.order.borrow()[0];
        let leader_pos = if leader == i { my_pos } else { self.solvers[leader].borrow().pos };
        (leader_pos - my_pos).max(0.0)
    }

    pub fn distance_to_ahead(&self, i: usize, my_pos: f64) -> Option<f64> {
        let r = self.rank_of.borrow()[i];
        if r <= 1 {
            return None;
        }
        let ahead = self.order.borrow()[r - 2];
        Some((self.solvers[ahead].borrow().pos - my_pos).max(0.0))
    }

    pub fn distance_to_behind(&self, i: usize, my_pos: f64) -> Option<f64> {
        let r = self.rank_of.borrow()[i];
        if r >= self.n {
            return None;
        }
        let behind = self.order.borrow()[r];
        Some((my_pos - self.solvers[behind].borrow().pos).max(0.0))
    }

    pub fn lane_gap_to_ahead(&self, i: usize, my_lane: f64) -> Option<f64> {
        let r = self.rank_of.borrow()[i];
        if r <= 1 {
            return None;
        }
        let ahead = self.order.borrow()[r - 2];
        Some(self.solvers[ahead].borrow().lane - my_lane)
    }

    pub fn lane_gap_to_behind(&self, i: usize, my_lane: f64) -> Option<f64> {
        let r = self.rank_of.borrow()[i];
        if r >= self.n {
            return None;
        }
        let behind = self.order.borrow()[r];
        Some(my_lane - self.solvers[behind].borrow().lane)
    }

    /// True on the tick this solver's rank improved (took over at least one
    /// other horse) versus the previous tick.
    pub fn did_overtake(&self, i: usize) -> bool {
        self.rank_of.borrow()[i] < self.prev_rank_of.borrow()[i]
    }

    /// Doc: front block = 0 < distGap < 2m AND |laneGap| <= a tolerance that
    /// shrinks from 0.75 horse-lanes at 0m to 0.3 at 2m. Lowest
    /// distance-gap qualifying horse wins. Returns (other_index, dist_gap).
    pub fn front_blocker(&self, i: usize, pos: f64, lane: f64) -> Option<(usize, f64)> {
        let mut best: Option<(usize, f64)> = None;
        for j in 0..self.n {
            if j == i {
                continue;
            }
            let other = self.solvers[j].borrow();
            let dist_gap = other.pos - pos;
            if !(dist_gap > 0.0 && dist_gap < 2.0) {
                continue;
            }
            let lane_gap = (other.lane - lane).abs();
            let tolerance = (1.0 - 0.6 * dist_gap / 2.0) * 0.75 * HORSE_LANE;
            if lane_gap <= tolerance && best.map_or(true, |(_, bd)| dist_gap < bd) {
                best = Some((j, dist_gap));
            }
        }
        best
    }

    /// Doc: side block = |distGap| < 1.05m AND |laneGap| < 2 horse-lanes.
    /// Returns (inner_blocked, outer_blocked) -- inner = lower lane (toward
    /// the rail), outer = higher lane.
    pub fn side_room(&self, i: usize, pos: f64, lane: f64) -> (bool, bool) {
        let mut inner_blocked = false;
        let mut outer_blocked = false;
        let lane_tol = 2.0 * HORSE_LANE;
        for j in 0..self.n {
            if j == i {
                continue;
            }
            let other = self.solvers[j].borrow();
            let d = other.pos - pos;
            if !(-1.05 < d && d < 1.05) {
                continue;
            }
            let lane_gap = other.lane - lane;
            if !(-lane_tol < lane_gap && lane_gap < lane_tol) {
                continue;
            }
            if lane_gap < 0.0 {
                inner_blocked = true;
                if outer_blocked {
                    break;
                }
            } else if lane_gap > 0.0 {
                outer_blocked = true;
                if inner_blocked {
                    break;
                }
            }
        }
        (inner_blocked, outer_blocked)
    }

    /// Doc: true if ALL THREE of Out/Front/Behind each find another horse
    /// (the same horse may satisfy more than one direction).
    pub fn is_surrounded(&self, i: usize, pos: f64, lane: f64) -> bool {
        let mut out = false;
        let mut front = false;
        let mut behind = false;
        for j in 0..self.n {
            if j == i {
                continue;
            }
            let other = self.solvers[j].borrow();
            let dist_gap = other.pos - pos;
            let lane_gap = other.lane - lane;
            if dist_gap.abs() < 1.5 && lane_gap > 0.0 && lane_gap < 3.0 * HORSE_LANE {
                out = true;
            }
            if dist_gap > 0.0 && dist_gap < 3.0 && lane_gap.abs() < 1.5 * HORSE_LANE {
                front = true;
            }
            if dist_gap > -3.0 && dist_gap < 0.0 && lane_gap.abs() < 1.5 * HORSE_LANE {
                behind = true;
            }
            if out && front && behind {
                break;
            }
        }
        out && front && behind
    }

    pub fn is_solo_front_runner(&self, i: usize) -> bool {
        for j in 0..self.n {
            if j == i {
                continue;
            }
            if strategy_matches(self.strategies[j], Strategy::Nige) {
                return false;
            }
        }
        true
    }

    /// Doc's pre-1.5-anniversary pacemaker rule: the 1st-place horse among
    /// the field's most-forward strategy group that's actually present.
    /// Depends only on `order` (refreshed once per tick by `update`) and
    /// solvers' fixed strategies, so it is computed once per tick and
    /// cached rather than recomputed per asking horse.
    pub fn pacemaker(&self) -> Option<usize> {
        if self.pacemaker_valid.get() {
            return self.pacemaker_idx.get();
        }
        let order = self.order.borrow();
        let mut result = None;
        for group in PACEMAKER_GROUPS {
            for &i in order.iter() {
                if group.contains(&self.strategies[i]) {
                    result = Some(i);
                    break;
                }
            }
            if result.is_some() {
                break;
            }
        }
        self.pacemaker_idx.set(result);
        self.pacemaker_valid.set(true);
        result
    }

    pub fn gap_to_pacemaker(&self, i: usize, my_pos: f64) -> f64 {
        match self.pacemaker() {
            Some(pm) if pm != i => self.solvers[pm].borrow().pos - my_pos,
            _ => 0.0,
        }
    }

    /// Doc: near = |distGap| < 3m AND |laneGap| < 3 horse-lanes.
    pub fn near_count(&self, i: usize, pos: f64, lane: f64) -> i32 {
        let lane_tol = 3.0 * HORSE_LANE;
        let mut count = 0;
        for j in 0..self.n {
            if j == i {
                continue;
            }
            let other = self.solvers[j].borrow();
            if (other.pos - pos).abs() < 3.0 && (other.lane - lane).abs() < lane_tol {
                count += 1;
            }
        }
        count
    }

    /// Doc: near_infront_count -- same "near" definition as near_count
    /// (|distGap|<3m, |laneGap|<3 horse-lanes) but restricted to horses AHEAD.
    pub fn near_infront_count(&self, i: usize, pos: f64, lane: f64) -> i32 {
        let lane_tol = 3.0 * HORSE_LANE;
        let mut count = 0;
        for j in 0..self.n {
            if j == i {
                continue;
            }
            let other = self.solvers[j].borrow();
            let dist_gap = other.pos - pos;
            if dist_gap > 0.0 && dist_gap < 3.0 && (other.lane - lane).abs() < lane_tol {
                count += 1;
            }
        }
        count
    }

    /// Doc: temptation_count_behind -- the number of horses ranked behind
    /// `i` (by finishing-position rank, not physical distance) that are
    /// currently rushing (kakari).
    pub fn temptation_count_behind(&self, i: usize) -> i32 {
        let r = self.rank_of.borrow()[i];
        let order = self.order.borrow();
        order[r..].iter().filter(|&&j| self.solvers[j].borrow().is_kakari).count() as i32
    }

    /// Doc: temptation_count_infront -- the number of horses ranked ahead
    /// of `i` that are currently rushing (kakari).
    pub fn temptation_count_infront(&self, i: usize) -> i32 {
        let r = self.rank_of.borrow()[i];
        let order = self.order.borrow();
        order[..r - 1].iter().filter(|&&j| self.solvers[j].borrow().is_kakari).count() as i32
    }

    /// Doc: running_style_temptation_count_{nige,senko,sashi,oikomi} -- the
    /// number of horses in the field (the querying horse counts too) with
    /// the given running strategy that are currently rushing. `self_idx`/
    /// `self_is_kakari` come from the caller directly (rather than
    /// re-borrowing `self.solvers[self_idx]`) since this is called from a
    /// condition check while that very solver is already mutably borrowed
    /// mid-`step()`.
    pub fn running_style_temptation_count(&self, strategy: Strategy, self_idx: usize, self_is_kakari: bool) -> i32 {
        let others = (0..self.n)
            .filter(|&j| j != self_idx && strategy_matches(self.strategies[j], strategy) && self.solvers[j].borrow().is_kakari)
            .count() as i32;
        let self_counts = strategy_matches(self.strategies[self_idx], strategy) && self_is_kakari;
        others + self_counts as i32
    }
}
